staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -301,14 +331,56 @@ func (w *watchCache) processEvent(event watch.Event, resourceVersion uint64, upd
 
 // Assumes that lock is already held for write.
 func (w *watchCache) updateCache(event *watchCacheEvent) {
-	if w.endIndex == w.startIndex+w.capacity {
+	w.resizeCache(event.RecordTime)
+	if w.isCacheFullLocked() {
 		// Cache is full - remove the oldest element.
 		w.startIndex++
 	}
 	w.cache[w.endIndex%w.capacity] = event
 	w.endIndex++
 }
 
+// Use dynamic size cache. If all cached events occur inside eventFreshDuration, increase cache capacity by 2x.
+// Decrease cache capacity by 2x when recent half events occur outside eventFreshDuration.
+func (w *watchCache) resizeCache(eventTime time.Time) {
+	if w.isCacheFullLocked() && eventTime.Sub(w.cache[w.startIndex%w.capacity].RecordTime) < eventFreshDuration {
+		capacity := min(w.capacity*2, w.upperBoundCapacity)
+		if capacity > w.capacity {
+			w.resizeCacheLocked(capacity)
+		}
+		return
+	}
+	if w.isCacheFullLocked() && eventTime.Sub(w.cache[(w.endIndex-w.capacity/2)%w.capacity].RecordTime) > eventFreshDuration {
```

Comment:
I missed that initially, but I don't want to resize by 2x if just half of the events fit into the duration.
Basically, we risk some flapping in that case.

We should change that to downsize by 2x, only if 1/4th of the events are ourside of eventFreshDuration. That should protect us from flapping.

Reply:
Done, PTAL.